redis_collections/lists.py

```diff
@@ -601,3 +637,270 @@ def sync_trans(pipe):
+
+        self.maxlen = maxlen
+        super(Deque, self).__init__(*args, **kwargs)
+
+    # Magic methods
+
+    def __delitem__(self, index):
+        if isinstance(index, slice):
+            raise TypeError
+
+        return super(Deque, self).__delitem__(index)
+
+    def __getitem__(self, index):
+        if isinstance(index, slice):
+            raise TypeError
+
+        return super(Deque, self).__getitem__(index)
+
+    def __setitem__(self, index, value):
+        if isinstance(index, slice):
+            raise TypeError
+
+        return super(Deque, self).__setitem__(index, value)
+
+    # Named methods
+
+    def _append_helper(self, value, pipe):
+        # Append on the right
+        len_self = pipe.rpush(self.key, self._pickle(value))
+        if self.writeback:
+            self.cache[len_self - 1] = value
+
+        # Check the length restriction
+        if (self.maxlen is None) or (len_self <= self.maxlen):
+            return
+
+        # Pop from the left
+        pipe.lpop(self.key)
+        if self.writeback:
+            items = six.iteritems(self.cache)
+            self.cache = {i - 1: v for i, v in items if i != 0}
+
+    def append(self, value):
+        """Add *value* to the right side of the collection."""
+        def append_trans(pipe):
+            self._append_helper(value, pipe)
+
+        self._transaction(append_trans)
+
+    def _appendleft_helper(self, value, pipe):
+        # Append on the left
+        len_self = pipe.lpush(self.key, self._pickle(value))
+        if self.writeback:
+            self.cache = {k + 1: v for k, v in six.iteritems(self.cache)}
+            self.cache[0] = value
+
+        # Check the length restriction
+        if (self.maxlen is None) or (len_self <= self.maxlen):
+            return
+
+        # Pop from the right
+        pipe.rpop(self.key)
+        if self.writeback:
+            cache_index = len_self - 1
+            items = six.iteritems(self.cache)
+            self.cache = {i: v for i, v in items if i != cache_index}
+
+    def appendleft(self, value):
+        """Add *value* to the left side of the collection."""
+        def appendleft_trans(pipe):
+            self._appendleft_helper(value, pipe)
+
+        self._transaction(appendleft_trans)
+
+    def copy(self, key=None):
+        """
+        Return a new :obj:`Deque` with the specified *key*. The new collection
+        will have the same values and maxlen as this collection.
+        """
+        other = self.__class__(
+            self.__iter__(),
+            self.maxlen,
+            redis=self.redis,
+            key=key,
+            writeback=self.writeback,
+        )
+
+        return other
+
+    def extend(self, other):
+        """
+        Extend the right side of the the collection by appending values from
+        the iterable *other*.
+        """
+        def extend_trans(pipe):
+            values = list(other.__iter__(pipe)) if use_redis else other
```

Comment:
It would be nice to only add the last `len(other) % self.maxlen` items if there is a size restriction, but I don't think the Python version does that.